From ec7389b6cff4d722833e0b9d79874fbb58806712 Mon Sep 17 00:00:00 2001
From: Steve Dickson <SteveD@redhat.com>
Date: Thu, 18 Sep 2014 14:20:00 -0400
Subject: [fs] xprtrdma: On disconnect, don't ignore pending CQEs

Message-id: <1411050057-12417-38-git-send-email-steved@redhat.com>
Patchwork-id: 96262
O-Subject: [RHEL7.1 PATCH 37/94] xprtrdma: On disconnect, don't ignore pending CQEs
Bugzilla: 1111708
RH-Acked-by: J. Bruce Fields <bfields@redhat.com>
RH-Acked-by: Benjamin Coddington <bcodding@redhat.com>

From: Chuck Lever <chuck.lever@oracle.com>

xprtrdma is currently throwing away queued completions during
a reconnect. RPC replies posted just before connection loss, or
successful completions that change the state of an FRMR, can be
missed.

Signed-off-by: Chuck Lever <chuck.lever@oracle.com>
Tested-by: Steve Wise <swise@opengridcomputing.com>
Tested-by: Shirley Ma <shirley.ma@oracle.com>
Tested-by: Devesh Sharma <devesh.sharma@emulex.com>
Signed-off-by: Anna Schumaker <Anna.Schumaker@Netapp.com>
Upstream Id: a7bc211ac926172ad20463afcf00ae7b9ebcd950
Bugzilla: 1111708
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/sunrpc/xprtrdma/verbs.c b/net/sunrpc/xprtrdma/verbs.c
index 13ff874..e49cdc9 100644
--- a/net/sunrpc/xprtrdma/verbs.c
+++ b/net/sunrpc/xprtrdma/verbs.c
@@ -310,6 +310,13 @@ rpcrdma_recvcq_upcall(struct ib_cq *cq, void *cq_context)
  rpcrdma_recvcq_poll(cq, ep);
 }
 
+static void
+rpcrdma_flush_cqs(struct rpcrdma_ep *ep)
+{
+ rpcrdma_recvcq_upcall(ep->rep_attr.recv_cq, ep);
+ rpcrdma_sendcq_upcall(ep->rep_attr.send_cq, ep);
+}
+
 #ifdef RPC_DEBUG
 static const char * const conn[] = {
  "address resolved",
@@ -872,9 +879,7 @@ retry:
   if (rc && rc != -ENOTCONN)
    dprintk("RPC:       %s: rpcrdma_ep_disconnect"
     " status %i\n", __func__, rc);
-
-  rpcrdma_clean_cq(ep->rep_attr.recv_cq);
-  rpcrdma_clean_cq(ep->rep_attr.send_cq);
+  rpcrdma_flush_cqs(ep);
 
   xprt = container_of(ia, struct rpcrdma_xprt, rx_ia);
   id = rpcrdma_create_id(xprt, ia,
@@ -985,8 +990,7 @@ rpcrdma_ep_disconnect(struct rpcrdma_ep *ep, struct rpcrdma_ia *ia)
 {
  int rc;
 
- rpcrdma_clean_cq(ep->rep_attr.recv_cq);
- rpcrdma_clean_cq(ep->rep_attr.send_cq);
+ rpcrdma_flush_cqs(ep);
  rc = rdma_disconnect(ia->ri_id);
  if (!rc) {
   /* returns without wait if not connected */
-- 
1.7.1